### PR TITLE
[AArch64] Separate PNR into its own Register Class

### PR DESCRIPTION
**Pull request by MDevereau**
This patch separates PNR registers into their own register class instead of sharing a register class with PPR registers. This primarily allows us to return more accurate register classes when applying assembly constraints, but also more protection from supplying an incorrect predicate type to an invalid register operand.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -3658,6 +3658,32 @@ void AArch64InstrInfo::copyPhysReg(MachineBasicBlock &MBB,
     return;
   }
 
+  if (AArch64::PNRRegClass.contains(DestReg) &&
+      AArch64::PPRRegClass.contains(SrcReg)) {
+    assert(Subtarget.hasSVEorSME() && "Unexpected SVE register.");
+    if ((DestReg.id() - AArch64::PN0) == (SrcReg.id() - AArch64::P0))
+      return;
+    DestReg = DestReg - AArch64::PN0 + AArch64::P0;
+    BuildMI(MBB, I, DL, get(AArch64::ORR_PPzPP), DestReg)
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:3667]: Should this also add an implicitDef of the corresponding predicate-as-counter register?

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64InstPrinter.cpp [hunk 2]
```diff
@@ -1348,9 +1348,14 @@ void AArch64InstPrinter::printPredicateAsCounter(const MCInst *MI,
                                                  const MCSubtargetInfo &STI,
                                                  raw_ostream &O) {
   unsigned Reg = MI->getOperand(OpNum).getReg();
+  O << "pn";
+  if (Reg >= AArch64::P0 && Reg <= AArch64::P15)
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/MCTargetDesc/AArch64InstPrinter.cpp:1352]: Is this change needed? I would expect a predicate-as-counter to be printed as a predicate-as-mask, but not the other way around.

Comment (MDevereau) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64InstPrinter.cpp:1352]: Yeah you're right, It's not needed after removing the PPRPNR classes from earlier. I've simplified this to 
```c++
if (Reg < AArch64::PN0 || Reg > AArch64::PN15)
    llvm_unreachable("Unsupported predicate-as-counter register");
O << "pn" << Reg - AArch64::PN0;
```

**Review by david-arm — COMMENTED** [review]
I've not done a full review, but I've left the comments I have so far!

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 3]
```diff
@@ -3658,6 +3658,32 @@ void AArch64InstrInfo::copyPhysReg(MachineBasicBlock &MBB,
     return;
   }
 
+  if (AArch64::PNRRegClass.contains(DestReg) &&
+      AArch64::PPRRegClass.contains(SrcReg)) {
+    assert(Subtarget.hasSVEorSME() && "Unexpected SVE register.");
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp (hunk 1 above)
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Yes it should, I've added it

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 4]
```diff
@@ -3658,6 +3658,32 @@ void AArch64InstrInfo::copyPhysReg(MachineBasicBlock &MBB,
...
+        .addReg(SrcReg, getKillRegState(KillSrc));
+    return;
+  }
+
+  if (AArch64::PPRRegClass.contains(DestReg) &&
+      AArch64::PNRRegClass.contains(SrcReg)) {
+    assert(Subtarget.hasSVEorSME() && "Unexpected SVE register.");
+    if ((DestReg.id() - AArch64::P0) == (SrcReg.id() - AArch64::PN0))
+      return;
+    SrcReg = (SrcReg - AArch64::PN0) + AArch64::P0;
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 5]
```diff
@@ -3658,6 +3658,32 @@ void AArch64InstrInfo::copyPhysReg(MachineBasicBlock &MBB,
...
+    BuildMI(MBB, I, DL, get(AArch64::ORR_PPzPP), DestReg)
+        .addReg(SrcReg)
+        .addReg(SrcReg)
+        .addReg(SrcReg, getKillRegState(KillSrc));
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64InstPrinter.cpp (hunk 2 above)
Comment (MDevereau) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64InstPrinter.cpp]: No it's not needed, I've removed the normal predicate logic and tidied this if block up a bit.

File: llvm/test/CodeGen/AArch64/spillfill-sve.mir [hunk 6]
```diff
@@ -61,6 +63,49 @@ body:             |
...
+  bb.0.entry:
+    liveins: $pn0
+
+    ; CHECK-LABEL: name: spills_fills_stack_id_pnr
+    ; CHECK: stack:
+    ; CHECK:      - { id: 0, name: '', type: spill-slot, offset: 0, size: 2, alignment: 2
+    ; CHECK-NEXT:     stack-id: scalable-vector, callee-saved-register: ''
+
+    ; EXPAND-LABEL: name: spills_fills_stack_id_pnr
+    ; EXPAND: STR_PXI $pn0, $sp, 7
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/CodeGen/AArch64/spillfill-sve.mir]: I've added some changes to the spill/fill logic and this test is now outputting and asserting `STR_PXI $p0, $sp, 7,implicit-def $pn0 `

File: llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir [hunk 7]
```diff
@@ -0,0 +1,58 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir:1]: The logic was they're both from organic examples. For example in `sve2p1-intrinsics-predicate-as-counter.ll` the pext tests will generate a PPR->PNR copy while the ptrue tests in that file will generate a PNR->PPR copy. If you'd rather I combine them into a single file and make them near carbon copies of each other that's fine by me.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 8]
```diff
@@ -3993,6 +4000,10 @@ void AArch64InstrInfo::storeRegToStackSlot(MachineBasicBlock &MBB,
     } else if (AArch64::PNRRegClass.hasSubClassEq(RC)) {
       assert((Subtarget.hasSVE2p1() || Subtarget.hasSME2()) &&
              "Unexpected register store without SVE2p1 or SME2");
+      if (SrcReg != AArch64::SP) {
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: I saw it done elsewhere in this function and wanted to be conservative, but yes, it is pointless. Removed.

File: llvm/test/CodeGen/AArch64/PNRtoPPRCopy.mir [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+frameInfo:
+  maxAlignment:    1
+  maxCallFrameSize: 0
+machineFunctionInfo:
+  hasRedZone:      false
+body:             |
+  bb.0 (%ir-block.0):
+    ; CHECK-LABEL: name: ptrue_d
+    ; CHECK: renamable $pn8 = PTRUE_C_D
+    ; CHECK-NEXT: $p0 = ORR_PPzPP $p8, $p8, killed $p8, implicit-def $pn8
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/CodeGen/AArch64/PNRtoPPRCopy.mir]: I've changed the implicit def to be the destination register instead of the source now.

File: llvm/test/CodeGen/AArch64/spillfill-sve.mir [hunk 10]
```diff
@@ -80,8 +80,8 @@ body:             |
     ; CHECK-NEXT:     stack-id: scalable-vector, callee-saved-register: ''
 
     ; EXPAND-LABEL: name: spills_fills_stack_id_pnr
-    ; EXPAND: STR_PXI $pn0, $sp, 7
-    ; EXPAND: $pn0 = LDR_PXI $sp, 7
+    ; EXPAND: STR_PXI $p0, $sp, 7, implicit-def $pn0
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/CodeGen/AArch64/spillfill-sve.mir]: I've removed the implicit def from store spills.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 11]
```diff
@@ -4156,6 +4170,10 @@ void AArch64InstrInfo::loadRegFromStackSlot(MachineBasicBlock &MBB,
     } else if (AArch64::PNRRegClass.hasSubClassEq(RC)) {
       assert((Subtarget.hasSVE2p1() || Subtarget.hasSME2()) &&
              "Unexpected register load without SVE2p1 or SME2");
+      if (DestReg != AArch64::SP) {
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: I saw it done elsewhere in this function and wanted to be conservative, but yes, it is pointless. Removed.

File: llvm/lib/Target/AArch64/AArch64RegisterInfo.td [hunk 12]
```diff
@@ -909,69 +929,89 @@ class PPRAsmOperand <string name, string RegClass, int Width>: AsmOperandClass {
...
-def PPRAsmOp32  : PPRAsmOperand<"PredicateS",   "PPR", 32>;
-def PPRAsmOp64  : PPRAsmOperand<"PredicateD",   "PPR", 64>;
-
-def PPRAny : PPRRegOp<"",  PPRAsmOpAny, ElementSizeNone, PPR>;
-def PPR8   : PPRRegOp<"b", PPRAsmOp8,   ElementSizeB,  PPR>;
-def PPR16  : PPRRegOp<"h", PPRAsmOp16,  ElementSizeH,  PPR>;
-def PPR32  : PPRRegOp<"s", PPRAsmOp32,  ElementSizeS,  PPR>;
-def PPR64  : PPRRegOp<"d", PPRAsmOp64,  ElementSizeD,  PPR>;
-
+def PPRAsmOpAny   : PPRAsmOperand<"PredicateAny",   "PPR",     0>;
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64RegisterInfo.td:932]: Okay, if it's not used at the moment, please remove it from this patch. If it's needed in the future we can always add it later.

File: llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir (hunk 7 above)
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir:1]: > If you'd rather I combine them into a single file and make them near carbon copies of each other that's fine by me
That would have my preference, thanks!

File: llvm/lib/Target/AArch64/AArch64RegisterInfo.td (hunk 12 above)
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64RegisterInfo.td:932]: Sorry, just ignore my comment above. The code is already in and used in `getPredicateRegisterClass` for inline-asm constraints. I'm not sure if anyone will ever use this for mask-predicates, but I guess there's no need to remove that functionality.

File: llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir (hunk 7 above)
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/CodeGen/AArch64/PPRtoPNRCopy.mir]: I've combined and simplified them into PNRtoPPRCopy.mir now.

File: llvm/test/CodeGen/AArch64/PNRtoPPRCopy.mir [hunk 13]
```diff
@@ -23,8 +11,8 @@ frameInfo:
 machineFunctionInfo:
   hasRedZone:      false
 body:             |
-  bb.0 (%ir-block.0):
-    ; CHECK-LABEL: name: ptrue_d
+  bb.0:
+    ; CHECK-LABEL: name: pnr_to_ppr
     ; CHECK: renamable $pn8 = PTRUE_C_D
     ; CHECK-NEXT: $p0 = ORR_PPzPP $p8, $p8, killed $p8, implicit-def $pn0
```
Comment (sdesmalen-arm) [inline llvm/test/CodeGen/AArch64/PNRtoPPRCopy.mir:17]: Should it do an `implicit-def` of `$pn0` here? It's copying `$pn8` -> `$p0`.
I would expect there only to be an `implicit-def` for a `$pn*` register if the destination of the COPY is a predicate-as-counter register.

**Review by sdesmalen-arm — APPROVED** [review]
LGTM
